README.md
# Motication of Fork
https://github.com/abema/abema-web/issues/15269

# polyfill.io User Agent normaliser

This is the VCL & JS implementation of the User-Agent normalisation used within the [polyfill-service](https://github.com/Financial-Times/polyfill-service) and the [polyfill-library](https://github.com/Financial-Times/polyfill-library).

# Usage

For a request which was the user-agent Chrome 71.1:

# Example in VCL
```vcl
import "normalise-user-agent.vcl";

sub vcl_recv {
  call normalise_user_agent_1_0_6;
#   req.http.normalized_user_agent_family = "chrome";
#   req.http.normalized_user_agent_major_version = "71";
#   req.http.normalized_user_agent_minor_version = "1";
#   req.http.normalized_user_agent_patch_version = "0";
#   req.http.Normalized-User-Agent = "chrome/71.1.0";
}
```

# Example in JS

```js
const UA = require('@financial-times/polyfill-useragent-normaliser');

const useragent = new UA(request.headers['user-agent']);
console.log(useragent.isUnknown()); // false
console.log(useragent.satisfies("<50")); // false
console.log(useragent.satisfies(">50")); // true
```


CHANGELOG.md
# [1.7.0](https://github.com/Financial-Times/polyfill-useragent-normaliser/compare/v1.6.3...v1.7.0) (2020-03-19)


### Bug Fixes

* **deps:** bump semver from 7.1.1 to 7.1.3 ([42eec6f](https://github.com/Financial-Times/polyfill-useragent-normaliser/commit/42eec6f2a345ae0103f13bd299b798e81221b953))


### Features

* update to 1.5.1 of @financial-times/useragent_parser ([e2859cc](https://github.com/Financial-Times/polyfill-useragent-normaliser/commit/e2859cc40a80359d6b1e6b7af1a918507f3b9f9e))

## [1.6.3](https://github.com/Financial-Times/polyfill-useragent-normaliser/compare/v1.6.2...v1.6.3) (2020-01-15)


### Bug Fixes

* **deps:** bump semver from 5.6.0 to 7.1.1 ([42cae2c](https://github.com/Financial-Times/polyfill-useragent-normaliser/commit/42cae2cf5e2b854e89c50bdc2a0c680d9f5ae698))

## [1.6.2](https://github.com/Financial-Times/polyfill-useragent-normaliser/compare/v1.6.1...v1.6.2) (2020-01-15)


### Bug Fixes

* **deps:** bump @financial-times/useragent_parser from 1.2.1 to 1.4.1 ([bfde311](https://github.com/Financial-Times/polyfill-useragent-normaliser/commit/bfde3111fc821b6e5eb5660ad67640e8ee9e7058))

## [1.6.1](https://github.com/Financial-Times/polyfill-useragent-normaliser/compare/v1.6.0...v1.6.1) (2020-01-15)


### Bug Fixes

* **deps:** [security] bump npm from 6.10.1 to 6.13.6 ([f88e774](https://github.com/Financial-Times/polyfill-useragent-normaliser/commit/f88e7743fef97731cd23719b00c8a44e2cd66304))

# [1.6.0](https://github.com/Financial-Times/polyfill-useragent-normaliser/compare/v1.5.1...v1.6.0) (2020-01-13)


### Features

* **alias:** Continue mapping opera to chrome ([db37102](https://github.com/Financial-Times/polyfill-useragent-normaliser/commit/db37102))

## [1.5.1](https://github.com/Financial-Times/polyfill-useragent-normaliser/compare/v1.5.0...v1.5.1) (2020-01-13)


### Bug Fixes

* Ensure useragent family name is checked for every alias ([54ae910](https://github.com/Financial-Times/polyfill-useragent-normaliser/commit/54ae910))

# [1.5.0](https://github.com/Financial-Times/polyfill-useragent-normaliser/compare/v1.4.2...v1.5.0) (2019-11-15)


### Features

* **alias:** chromium-edge is now chrome ([3ae1eaf](https://github.com/Financial-Times/polyfill-useragent-normaliser/commit/3ae1eaf))

## [1.4.2](https://github.com/Financial-Times/polyfill-useragent-normaliser/compare/v1.4.1...v1.4.2) (2019-07-23)


### Bug Fixes

* add a description to package.json ([c934413](https://github.com/Financial-Times/polyfill-useragent-normaliser/commit/c934413))

## [1.4.1](https://github.com/Financial-Times/polyfill-useragent-normaliser/compare/v1.4.0...v1.4.1) (2019-07-23)


### Bug Fixes

* Automatically build new copy of library when the version changes ([78eb53b](https://github.com/Financial-Times/polyfill-useragent-normaliser/commit/78eb53b))

# [1.4.0](https://github.com/Financial-Times/polyfill-useragent-normaliser/compare/v1.3.0...v1.4.0) (2019-07-19)


### Bug Fixes

* add the commitlint config ([e4f5943](https://github.com/Financial-Times/polyfill-useragent-normaliser/commit/e4f5943))
* run prettier and use "use strict" ([7e53d7c](https://github.com/Financial-Times/polyfill-useragent-normaliser/commit/7e53d7c))
* update package-lock.json ([5a8d363](https://github.com/Financial-Times/polyfill-useragent-normaliser/commit/5a8d363))


### Features

* automate the releases using semantic release ([350d18f](https://github.com/Financial-Times/polyfill-useragent-normaliser/commit/350d18f))


compilers/fastly.js
"use strict";

const fs = require("fs");
const path = require("path");
const data = JSON.parse(
  fs.readFileSync(path.join(__dirname, "../data.json"), "utf8")
);
const version = JSON.parse(
  fs.readFileSync(path.join(__dirname, "../package.json"))
).version.replace(/\./g, "_");
const ua_parser_vcl = fs.readFileSync(
  require.resolve("@financial-times/useragent_parser/lib/ua_parser.vcl"),
  "utf8"
);

// Fastly uses strings as regex and does not want it to start or end with /
function removeRegexLiteralMarkers(regex) {
  return regex.substring(1, regex.length - 1);
}
let file = `${ua_parser_vcl}

sub normalise_user_agent_${version} {
  if (!req.http.User-Agent) {
    set req.http.normalized_user_agent_family = "other";
    set req.http.normalized_user_agent_major_version = "0";
    set req.http.normalized_user_agent_minor_version = "0";
    set req.http.normalized_user_agent_patch_version = "0";
  } else {
    if (req.http.User-Agent ~ {"${removeRegexLiteralMarkers(
      data.isNormalized
    )}"}) {
      set req.http.normalized_user_agent_family = std.tolower(re.group.1);
      set req.http.normalized_user_agent_major_version = re.group.2;
      set req.http.normalized_user_agent_minor_version = if(re.group.3, std.strtol(re.group.3, 10), "0");
      set req.http.normalized_user_agent_patch_version = "0";
    } else {
`;
// 1. Add the normalisations
for (const { reason, regex } of data.normalisations) {
  let s = "";
  s += `\n\t\t# ${reason}`;
  s += `\n\t\tset req.http.User-Agent = regsub(req.http.User-Agent, {"${removeRegexLiteralMarkers(
    regex
  )}"}, "");\n`;
  file += s;
}

// 2. Do the useragent parsing into family/major.minor.patch
file += `
        call useragent_parser;

        # Clone the original values for later modification. This helps when debugging as it let's us see what the useragent_parser function returned.
        set req.http.normalized_user_agent_family = req.http.useragent_parser_family;
        set req.http.normalized_user_agent_major_version = req.http.useragent_parser_major;
        set req.http.normalized_user_agent_minor_version = if(req.http.useragent_parser_minor != "", std.strtol(req.http.useragent_parser_minor, 10), "0");
`;

// For improved CDN cache performance, remove the patch version.  There are few cases in which a patch release drops the requirement for a polyfill, but if so, the polyfill can simply be served unnecessarily to the patch versions that contain the fix, and we can stop targeting at the next minor release.
file += `\n\t\tset req.http.normalized_user_agent_patch_version = "0";\n`;

// 3. Aliases
file += `\n\t\tset req.http.normalized_user_agent_family = std.tolower(req.http.useragent_parser_family);\n\t}`;
for (const [family, alias] of Object.entries(data.aliases)) {
  if (typeof alias === "string") {
    file += `\n\t\tif (req.http.normalized_user_agent_family == "${family}") {
            set req.http.normalized_user_agent_family = "${alias}";
		}`;
  } else if (Array.isArray(alias)) {
    file += `\n\t\tif (req.http.normalized_user_agent_family == "${family}") {
            set req.http.normalized_user_agent_family = "${alias[0]}";
            set req.http.normalized_user_agent_major_version = "${alias[1]}";
            set req.http.normalized_user_agent_minor_version = "0";
        }`;
  } else if (typeof alias === "object") {
    file += `\n\t\tif (req.http.normalized_user_agent_family == "${family}") {`;
    for (const [range, replacement] of Object.entries(alias)) {
      const [major, minor] = range.split(".");
      if (minor !== undefined) {
        file += `\n\t\t\tif (req.http.normalized_user_agent_family == "${family}" && req.http.normalized_user_agent_major_version == "${major}" && req.http.normalized_user_agent_minor_version == "${minor}") {`;
      } else {
        file += `\n\t\t\tif (req.http.normalized_user_agent_family == "${family}" && req.http.normalized_user_agent_major_version == "${major}") {`;
      }
      file += `
                set req.http.normalized_user_agent_family = "${replacement[0]}";
                set req.http.normalized_user_agent_major_version = "${replacement[1]}";
                set req.http.normalized_user_agent_minor_version = "0";
            }`;
    }
    file += `\n\t\t}`;
  }
}

// 4. Check if browser and version are in the baseline supported browser versions
file += `\n\t\tset req.http.normalized_user_agent_temp_version = req.http.normalized_user_agent_major_version req.http.normalized_user_agent_minor_version;`;
file += `\n\t\tif (\n            !req.http.Host`;
for (const [family, range] of Object.entries(data.baselineVersions)) {
  if (range === "*") {
    file += ` || \n            (req.http.normalized_user_agent_family == "${family}")`;
  } else {
    if (Number.isInteger(Number(range))) {
      file += ` || \n            (req.http.normalized_user_agent_family == "${family}" && std.atoi(req.http.normalized_user_agent_major_version) >= ${range})`;
    } else {
      const [major, minor] = range.split(".");
      file += ` || \n            (req.http.normalized_user_agent_family == "${family}" && std.atoi(req.http.normalized_user_agent_temp_version) >= ${major +
        minor})`;
    }
  }
}
file += `\n		) {} else {
            set req.http.normalized_user_agent_family = "other";
            set req.http.normalized_user_agent_major_version = "0";
            set req.http.normalized_user_agent_minor_version = "0";
        }`;
file += `\n\t\tunset req.http.normalized_user_agent_temp_version;`;
file += `
    }
    set req.http.Normalized-User-Agent = req.http.normalized_user_agent_family "/"  req.http.normalized_user_agent_major_version "." req.http.normalized_user_agent_minor_version "." req.http.normalized_user_agent_patch_version;
}`;

fs.writeFileSync(
  path.join(__dirname, "../lib/normalise-user-agent.vcl"),
  file,
  "utf8"
);

fs.writeFileSync(
  path.join(__dirname, "../fastly/vcl/main.vcl"),
  `
include "normalise-user-agent.vcl";

sub vcl_recv {
	error 903 "Normalise User Agent";
}

sub vcl_error {
	if (obj.status == 903) {
        call normalise_user_agent_${version};
		set obj.status = 200;
		set obj.response = "OK";
		set obj.http.Content-Type = "text/plain; charset=utf-8";
        set obj.http.normalized_user_agent_family = req.http.normalized_user_agent_family;
        set obj.http.normalized_user_agent_major_version = req.http.normalized_user_agent_major_version;
        set obj.http.normalized_user_agent_minor_version = req.http.normalized_user_agent_minor_version;
        set obj.http.normalized_user_agent_patch_version = req.http.normalized_user_agent_patch_version;
		set obj.http.Normalized-User-Agent = req.http.Normalized-User-Agent;
		synthetic req.http.Normalized-User-Agent;
		return (deliver);
	}
}`,
  "utf8"
);


commitlint.config.js
"use strict";

module.exports = {
  rules: {
    "body-leading-blank": [1, "always"],
    "footer-leading-blank": [1, "always"],
    "header-max-length": [2, "always", Infinity],
    "scope-case": [2, "always", "lower-case"],
    "subject-case": [
      2,
      "always",
      [
        "lower-case", // default
        "upper-case", // UPPERCASE
        "camel-case", // camelCase
        "kebab-case", // kebab-case
        "pascal-case", // PascalCase
        "sentence-case", // Sentence case
        "snake-case", // snake_case
        "start-case" // Start Case
      ]
    ],
    "subject-empty": [2, "never"],
    "subject-full-stop": [0, "never", "."],
    "type-case": [2, "always", "lower-case"],
    "type-empty": [2, "never"],
    "type-enum": [
      2,
      "always",
      [
        "build",
        "chore",
        "ci",
        "docs",
        "feat",
        "fix",
        "perf",
        "refactor",
        "revert",
        "style",
        "test"
      ]
    ]
  }
};


lib/normalise-user-agent.js
"use strict";

const useragent = require("@financial-times/useragent_parser");
const semver = require("semver");

function UA(uaString) {
  let normalized;
  if (!uaString) {
    this.ua = {};
    this.ua.family = "other";
    this.ua.major = "0";
    this.ua.minor = "0";
    this.ua.patch = "0";
  } else {
    if (
      (normalized = uaString.match(/^(\w+)\/(\d+)(?:\.(\d+)(?:\.(\d+))?)?$/i))
    ) {
      this.ua = {
        family: normalized[1].toLowerCase(),
        major: normalized[2],
        minor: normalized[3] || "0",
        patch: "0"
      };
    } else {
      // Google Search iOS app should be detected as the underlying browser, which is safari on iOS
      uaString = uaString.replace(/ GSA\/[\d\.]+/i, "");

      // Instagram should be detected as the underlying browser, which is safari on ios
      uaString = uaString.replace(/ Instagram [\d\.]+/i, "");

      // WebPageTest is not a real browser, remove the token to find the underlying browser
      uaString = uaString.replace(/ PTST\/[\d\.]+/i, "");

      // Waterfox is a Firefox fork, we can remove the Waterfox identifiers and parse the result as Firefox
      uaString = uaString.replace(/ Waterfox\/[\d\.]+/i, "");

      // Pale Moon has a Firefox-compat UA string, we can remove the Pale Moon and Goanna identifiers and parse the result as Firefox
      uaString = uaString.replace(/ Goanna\/[\d\.]+/i, "");

      // Pale Moon has a Firefox-compat UA string, we can remove the Pale Moon and Goanna identifiers and parse the result as Firefox
      uaString = uaString.replace(/ PaleMoon\/[\d\.]+/i, "");

      // Yandex browser is recognised by UA module but is actually Chromium under the hood, so better to remove the Yandex identifier and get the UA module to detect it as Chrome
      uaString = uaString.replace(/(YaBrowser)\/(\d+\.)+\d+ /, "");

      // Crosswalk browser is recognised by UA module but is actually Chromium under the hood, so better to remove the identifier and get the UA module to detect it as Chrome
      uaString = uaString.replace(
        / (Crosswalk)\/(\d+)\.(\d+)\.(\d+)\.(\d+)/i,
        ""
      );

      // Chrome and Opera on iOS uses a UIWebView of the underlying platform to render content. By stripping the CriOS or OPiOS strings, the useragent parser will alias the user agent to ios_saf for the UIWebView, which is closer to the actual renderer
      uaString = uaString.replace(
        /((CriOS|OPiOS)\/(\d+)\.(\d+)\.(\d+)\.(\d+)|(FxiOS\/(\d+)\.(\d+)))/,
        ""
      );

      // Vivaldi browser is recognised by UA module but is actually identical to Chrome, so the best way to get accurate targeting is to remove the vivaldi token from the UA
      uaString = uaString.replace(/ vivaldi\/[\d\.]+\d+/i, "");

      // Facebook in-app browser `[FBAN/.....]` or `[FB_IAB/.....]` (see https://github.com/Financial-Times/polyfill-service/issues/990)
      uaString = uaString.replace(/ \[(FB_IAB|FBAN|FBIOS|FB4A)\/[^\]]+\]/i, "");

      // Electron/X.Y.Z` (see https://github.com/Financial-Times/polyfill-service/issues/1129)
      uaString = uaString.replace(/ Electron\/[\d\.]+\d+/i, "");

      // Chromium-based Edge
      uaString = uaString.replace(/ Edg\/[\d\.]+\d+/i, "");

      this.ua = useragent(uaString);

      this.ua.patch = "0";

      this.ua.family = this.ua.family.toLowerCase();
    }
    if (this.ua.family === "blackberry webkit") {
      this.ua.family = "bb";
    }
    if (this.ua.family === "blackberry") {
      this.ua.family = "bb";
    }
    if (this.ua.family === "pale moon (firefox variant)") {
      this.ua.family = "firefox";
    }
    if (this.ua.family === "pale moon") {
      this.ua.family = "firefox";
    }
    if (this.ua.family === "firefox mobile") {
      this.ua.family = "firefox_mob";
    }
    if (this.ua.family === "firefox namoroka") {
      this.ua.family = "firefox";
    }
    if (this.ua.family === "firefox shiretoko") {
      this.ua.family = "firefox";
    }
    if (this.ua.family === "firefox minefield") {
      this.ua.family = "firefox";
    }
    if (this.ua.family === "firefox alpha") {
      this.ua.family = "firefox";
    }
    if (this.ua.family === "firefox beta") {
      this.ua.family = "firefox";
    }
    if (this.ua.family === "microb") {
      this.ua.family = "firefox";
    }
    if (this.ua.family === "mozilladeveloperpreview") {
      this.ua.family = "firefox";
    }
    if (this.ua.family === "iceweasel") {
      this.ua.family = "firefox";
    }
    if (this.ua.family === "opera tablet") {
      this.ua.family = "opera";
    }
    if (this.ua.family === "opera mobile") {
      this.ua.family = "op_mob";
    }
    if (this.ua.family === "opera mini") {
      this.ua.family = "op_mini";
    }
    if (this.ua.family === "chrome mobile webview") {
      this.ua.family = "chrome";
    }
    if (this.ua.family === "chrome mobile") {
      this.ua.family = "chrome";
    }
    if (this.ua.family === "chrome frame") {
      this.ua.family = "chrome";
    }
    if (this.ua.family === "chromium") {
      this.ua.family = "chrome";
    }
    if (this.ua.family === "headlesschrome") {
      this.ua.family = "chrome";
    }
    if (this.ua.family === "ie mobile") {
      this.ua.family = "ie_mob";
    }
    if (this.ua.family === "ie large screen") {
      this.ua.family = "ie";
    }
    if (this.ua.family === "internet explorer") {
      this.ua.family = "ie";
    }
    if (this.ua.family === "edge mobile") {
      this.ua.family = "edge_mob";
    }
    if (this.ua.family === "uc browser") {
      if (
        this.ua.family === "uc browser" &&
        this.ua.major === "9" &&
        this.ua.minor === "9"
      ) {
        this.ua.family = "ie";
        this.ua.major = "10";
        this.ua.minor = "0";
      }
    }
    if (this.ua.family === "chrome mobile ios") {
      this.ua.family = "ios_chr";
    }
    if (this.ua.family === "mobile safari") {
      this.ua.family = "ios_saf";
    }
    if (this.ua.family === "iphone") {
      this.ua.family = "ios_saf";
    }
    if (this.ua.family === "iphone simulator") {
      this.ua.family = "ios_saf";
    }
    if (this.ua.family === "mobile safari uiwebview") {
      this.ua.family = "ios_saf";
    }
    if (this.ua.family === "mobile safari ui/wkwebview") {
      this.ua.family = "ios_saf";
    }
    if (this.ua.family === "samsung internet") {
      this.ua.family = "samsung_mob";
    }
    if (this.ua.family === "phantomjs") {
      this.ua.family = "safari";
      this.ua.major = "5";
      this.ua.minor = "0";
    }
    if (this.ua.family === "opera") {
      if (this.ua.family === "opera" && this.ua.major === "20") {
        this.ua.family = "chrome";
        this.ua.major = "33";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "21") {
        this.ua.family = "chrome";
        this.ua.major = "34";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "22") {
        this.ua.family = "chrome";
        this.ua.major = "35";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "23") {
        this.ua.family = "chrome";
        this.ua.major = "36";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "24") {
        this.ua.family = "chrome";
        this.ua.major = "37";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "25") {
        this.ua.family = "chrome";
        this.ua.major = "38";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "26") {
        this.ua.family = "chrome";
        this.ua.major = "39";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "27") {
        this.ua.family = "chrome";
        this.ua.major = "40";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "28") {
        this.ua.family = "chrome";
        this.ua.major = "41";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "29") {
        this.ua.family = "chrome";
        this.ua.major = "42";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "30") {
        this.ua.family = "chrome";
        this.ua.major = "43";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "31") {
        this.ua.family = "chrome";
        this.ua.major = "44";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "32") {
        this.ua.family = "chrome";
        this.ua.major = "45";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "33") {
        this.ua.family = "chrome";
        this.ua.major = "46";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "34") {
        this.ua.family = "chrome";
        this.ua.major = "47";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "35") {
        this.ua.family = "chrome";
        this.ua.major = "48";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "36") {
        this.ua.family = "chrome";
        this.ua.major = "49";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "37") {
        this.ua.family = "chrome";
        this.ua.major = "50";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "38") {
        this.ua.family = "chrome";
        this.ua.major = "51";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "39") {
        this.ua.family = "chrome";
        this.ua.major = "52";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "40") {
        this.ua.family = "chrome";
        this.ua.major = "53";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "41") {
        this.ua.family = "chrome";
        this.ua.major = "54";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "42") {
        this.ua.family = "chrome";
        this.ua.major = "55";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "43") {
        this.ua.family = "chrome";
        this.ua.major = "56";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "44") {
        this.ua.family = "chrome";
        this.ua.major = "57";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "45") {
        this.ua.family = "chrome";
        this.ua.major = "58";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "46") {
        this.ua.family = "chrome";
        this.ua.major = "59";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "47") {
        this.ua.family = "chrome";
        this.ua.major = "60";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "48") {
        this.ua.family = "chrome";
        this.ua.major = "61";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "49") {
        this.ua.family = "chrome";
        this.ua.major = "62";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "50") {
        this.ua.family = "chrome";
        this.ua.major = "63";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "51") {
        this.ua.family = "chrome";
        this.ua.major = "64";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "52") {
        this.ua.family = "chrome";
        this.ua.major = "65";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "53") {
        this.ua.family = "chrome";
        this.ua.major = "66";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "54") {
        this.ua.family = "chrome";
        this.ua.major = "67";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "55") {
        this.ua.family = "chrome";
        this.ua.major = "68";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "56") {
        this.ua.family = "chrome";
        this.ua.major = "69";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "57") {
        this.ua.family = "chrome";
        this.ua.major = "70";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "58") {
        this.ua.family = "chrome";
        this.ua.major = "71";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "59") {
        this.ua.family = "chrome";
        this.ua.major = "72";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "60") {
        this.ua.family = "chrome";
        this.ua.major = "73";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "61") {
        this.ua.family = "chrome";
        this.ua.major = "74";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "62") {
        this.ua.family = "chrome";
        this.ua.major = "75";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "63") {
        this.ua.family = "chrome";
        this.ua.major = "76";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "64") {
        this.ua.family = "chrome";
        this.ua.major = "77";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "65") {
        this.ua.family = "chrome";
        this.ua.major = "78";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "66") {
        this.ua.family = "chrome";
        this.ua.major = "79";
        this.ua.minor = "0";
      }
      if (this.ua.family === "opera" && this.ua.major === "67") {
        this.ua.family = "chrome";
        this.ua.major = "80";
        this.ua.minor = "0";
      }
    }
    if (this.ua.family === "googlebot") {
      if (
        this.ua.family === "googlebot" &&
        this.ua.major === "2" &&
        this.ua.minor === "1"
      ) {
        this.ua.family = "chrome";
        this.ua.major = "41";
        this.ua.minor = "0";
      }
    }
    if (
      false ||
      this.ua.family === "edge" ||
      this.ua.family === "edge_mob" ||
      (this.ua.family === "ie" && Number(this.ua.major) >= 8) ||
      (this.ua.family === "ie_mob" && Number(this.ua.major) >= 11) ||
      (this.ua.family === "chrome" && Number(this.ua.major) >= 29) ||
      (this.ua.family === "safari" && Number(this.ua.major) >= 9) ||
      (this.ua.family === "ios_saf" && Number(this.ua.major) >= 9) ||
      (this.ua.family === "ios_chr" && Number(this.ua.major) >= 9) ||
      (this.ua.family === "firefox" && Number(this.ua.major) >= 38) ||
      (this.ua.family === "firefox_mob" && Number(this.ua.major) >= 38) ||
      (this.ua.family === "android" &&
        Number(this.ua.major + "." + this.ua.minor) >= 4.3) ||
      (this.ua.family === "opera" && Number(this.ua.major) >= 33) ||
      (this.ua.family === "op_mob" && Number(this.ua.major) >= 10) ||
      (this.ua.family === "op_mini" && Number(this.ua.major) >= 5) ||
      (this.ua.family === "bb" && Number(this.ua.major) >= 6) ||
      (this.ua.family === "samsung_mob" && Number(this.ua.major) >= 4)
    ) {
    } else {
      this.ua.family = "other";
      this.ua.major = "0";
      this.ua.minor = "0";
      this.ua.patch = "0";
    }
  }
  this.version =
    (Number(this.ua.major) || 0) + "." + (Number(this.ua.minor) || 0) + ".0";
}

UA.prototype.getFamily = function() {
  return this.ua.family;
};

UA.prototype.getVersion = function() {
  return this.version;
};

UA.prototype.satisfies = function(range) {
  return semver.satisfies(this.version, range);
};

UA.prototype.getBaseline = function() {
  return UA.getBaselines()[this.ua.family];
};

UA.prototype.meetsBaseline = function() {
  return semver.satisfies(
    this.version,
    ">=" + UA.getBaselines()[this.ua.family]
  );
};

UA.prototype.isUnknown = function() {
  return (
    Object.keys(UA.getBaselines()).indexOf(this.ua.family) === -1 ||
    !this.meetsBaseline()
  );
};

UA.normalize = function(uaString) {
  const ua = new UA(uaString);
  return ua.getFamily() + "/" + ua.version;
};

UA.getBaselines = function() {
  return {
    edge: "*",
    edge_mob: "*",
    ie: "8",
    ie_mob: "11",
    chrome: "29",
    safari: "9",
    ios_saf: "9",
    ios_chr: "9",
    firefox: "38",
    firefox_mob: "38",
    android: "4.3",
    opera: "33",
    op_mob: "10",
    op_mini: "5",
    bb: "6",
    samsung_mob: "4"
  };
};

module.exports = UA;
